File: frontend/src/api/aiFoundation.ts
import { supabase } from '../lib/supabase';
import { callEdgeFunction } from '../lib/edgeFunctions';
import type {
  AIProfile,
  TradeEvaluation,
  KnowledgeLink,
  DetectedPattern,
  CoachingSession,
  AIInsight,
  AIRecommendation,
  AISummary,
  AIProviderConfig,
  AIDashboardData,
  KnowledgeGraphData,
} from './types';

export const aiFoundationService = {
  // Profile
  getProfile: async (projectId: string): Promise<AIProfile> => {
    const { data, error } = await supabase
      .from('ai_profile')
      .select('*')
      .eq('project_id', projectId)
      .maybeSingle();
    if (error) throw error;
    return (data ?? {}) as AIProfile;
  },

  updateProfile: async (projectId: string, profile: Partial<AIProfile>): Promise<AIProfile> => {
    const { data, error } = await supabase
      .from('ai_profile')
      .upsert({ project_id: projectId, ...profile, updated_at: new Date().toISOString() }, { onConflict: 'project_id' })
      .select()
      .single();
    if (error) throw error;
    return data as AIProfile;
  },

  analyzeProfile: async (projectId: string): Promise<AIProfile> =>
    callEdgeFunction('ai', { operation: 'analyze-profile', project_id: projectId }),

  // Evaluations
  evaluateTrade: async (projectId: string, tradeId: string): Promise<TradeEvaluation> =>
    callEdgeFunction('ai', { operation: 'evaluate-trade', project_id: projectId, data: { trade_id: tradeId } }),

  getEvaluations: async (projectId: string, limit = 50): Promise<TradeEvaluation[]> => {
    const { data, error } = await supabase
      .from('trade_evaluation')
      .select('*')
      .eq('project_id', projectId)
      .order('created_at', { ascending: false })
      .limit(limit);
    if (error) throw error;
    return (data ?? []) as TradeEvaluation[];
  },

  // Patterns
  detectPatterns: async (projectId: string): Promise<DetectedPattern[]> =>
    callEdgeFunction('ai', { operation: 'detect-patterns', project_id: projectId }),

  getPatterns: async (projectId: string, patternType?: string): Promise<DetectedPattern[]> => {
    let query = supabase.from('detected_pattern').select('*').eq('project_id', projectId);
    if (patternType) query = query.eq('pattern_type', patternType);
    const { data, error } = await query.order('last_detected_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as DetectedPattern[];
  },

  // Knowledge Links
  createLink: async (projectId: string, link: { source_type: string; source_id: string; target_type: string; target_id: string; relationship: string }): Promise<KnowledgeLink> => {
    const { data, error } = await supabase
      .from('knowledge_link')
      .insert({ project_id: projectId, ...link })
      .select()
      .single();
    if (error) throw error;
    return data as KnowledgeLink;
  },

  getLinks: async (projectId: string, entityType?: string, entityId?: string): Promise<KnowledgeLink[]> => {
    let query = supabase.from('knowledge_link').select('*').eq('project_id', projectId);
    if (entityType && entityId) {
      query = query.or(
        `and(source_type.eq.${entityType},source_id.eq.${entityId}),and(target_type.eq.${entityType},target_id.eq.${entityId})`
      );
    }
    const { data, error } = await query.order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as KnowledgeLink[];
  },

  deleteLink: async (projectId: string, linkId: string): Promise<void> => {
    const { error } = await supabase
      .from('knowledge_link')
      .delete()
      .eq('id', linkId)
      .eq('project_id', projectId);
    if (error) throw error;
  },

  autoLink: async (projectId: string): Promise<{ linked: number }> =>
    callEdgeFunction('ai', { operation: 'auto-link', project_id: projectId }),

  getGraph: async (projectId: string, entityType?: string, entityId?: string): Promise<KnowledgeGraphData> =>
    callEdgeFunction('ai', { operation: 'knowledge-graph', project_id: projectId, data: { entity_type: entityType, entity_id: entityId } }),

  // Insights
  generateInsights: async (projectId: string): Promise<AIInsight[]> =>
    callEdgeFunction('ai', { operation: 'generate-insights', project_id: projectId }),

  getInsights: async (projectId: string): Promise<AIInsight[]> => {
    const { data, error } = await supabase
      .from('ai_insight')
      .select('*')
      .eq('project_id', projectId)
      .eq('is_dismissed', false)
      .order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as AIInsight[];
  },

  dismissInsight: async (projectId: string, insightId: string): Promise<void> => {
    const { error } = await supabase
      .from('ai_insight')
      .update({ is_dismissed: true })
      .eq('id', insightId)
      .eq('project_id', projectId);
    if (error) throw error;
  },

  // Recommendations
  generateRecommendations: async (projectId: string): Promise<AIRecommendation[]> =>
    callEdgeFunction('ai', { operation: 'generate-recommendations', project_id: projectId }),

  getRecommendations: async (projectId: string): Promise<AIRecommendation[]> => {
    const { data, error } = await supabase
      .from('ai_recommendation')
      .select('*')
      .eq('project_id', projectId)
      .eq('is_dismissed', false)
      .order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as AIRecommendation[];
  },

  dismissRecommendation: async (projectId: string, recId: string): Promise<void> => {
    const { error } = await supabase
      .from('ai_recommendation')
      .update({ is_dismissed: true })
      .eq('id', recId)
      .eq('project_id', projectId);
    if (error) throw error;
  },

  // Coaching
  generateCoaching: async (projectId: string, sessionType = 'daily', date?: string): Promise<CoachingSession> =>
    callEdgeFunction('ai', { operation: 'generate-coaching', project_id: projectId, data: { session_type: sessionType, date } }),

  getCoaching: async (projectId: string, sessionType?: string): Promise<CoachingSession[]> => {
    let query = supabase.from('coaching_session').select('*').eq('project_id', projectId);
    if (sessionType) query = query.eq('session_type', sessionType);
    const { data, error } = await query.order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as CoachingSession[];
  },

  // Summaries
  createSummary: async (projectId: string, summary: { summary_type: string; content?: Record<string, unknown>; text_summary?: string }): Promise<AISummary> => {
    const { data, error } = await supabase
      .from('ai_summary')
      .insert({ project_id: projectId, ...summary })
      .select()
      .single();
    if (error) throw error;
    return data as AISummary;
  },

  getSummaries: async (projectId: string, summaryType?: string, period?: string): Promise<AISummary[]> => {
    let query = supabase.from('ai_summary').select('*').eq('project_id', projectId);
    if (summaryType) query = query.eq('summary_type', summaryType);
    if (period) query = query.eq('period_start', period);
    const { data, error } = await query.order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as AISummary[];
  },

  generatePerformanceSummary: async (projectId: string): Promise<AISummary> =>
    callEdgeFunction('ai', { operation: 'generate-performance-summary', project_id: projectId }),

  // Context
  buildContext: async (projectId: string, options?: Record<string, unknown>): Promise<Record<string, unknown>> =>
    callEdgeFunction('ai', { operation: 'build-context', project_id: projectId, data: options ?? {} }),

  // Providers (no project_id)
  getProviders: async (): Promise<AIProviderConfig[]> => {
    const { data, error } = await supabase
      .from('ai_provider_config')
      .select('*')
      .order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as AIProviderConfig[];
  },

  getDefaultProvider: async (): Promise<AIProviderConfig> => {
    const { data, error } = await supabase
      .from('ai_provider_config')
      .select('*')
      .eq('is_default', true)
      .maybeSingle();
    if (error) throw error;
    return (data ?? {}) as AIProviderConfig;
  },

  createProvider: async (provider: Partial<AIProviderConfig>): Promise<AIProviderConfig> => {
    const { data, error } = await supabase
      .from('ai_provider_config')
      .insert(provider)
      .select()
      .single();
    if (error) throw error;
    return data as AIProviderConfig;
  },

  updateProvider: async (providerId: string, provider: Partial<AIProviderConfig>): Promise<AIProviderConfig> => {
    const { data, error } = await supabase
      .from('ai_provider_config')
      .update(provider)
      .eq('id', providerId)
      .select()
      .single();
    if (error) throw error;
    return data as AIProviderConfig;
  },

  deleteProvider: async (providerId: string): Promise<void> => {
    const { error } = await supabase
      .from('ai_provider_config')
      .delete()
      .eq('id', providerId);
    if (error) throw error;
  },

  // Dashboard
  getDashboard: async (projectId: string): Promise<AIDashboardData> => {
    const [profile, insights, coaching, recommendations, patterns] = await Promise.all([
      supabase.from('ai_profile').select('*').eq('project_id', projectId).maybeSingle(),
      supabase.from('ai_insight').select('*').eq('project_id', projectId).eq('is_dismissed', false).order('created_at', { ascending: false }).limit(5),
      supabase.from('coaching_session').select('*').eq('project_id', projectId).eq('is_read', false).order('created_at', { ascending: false }).limit(5),
      supabase.from('ai_recommendation').select('*').eq('project_id', projectId).eq('is_dismissed', false).order('created_at', { ascending: false }).limit(5),
      supabase.from('detected_pattern').select('*').eq('project_id', projectId).eq('is_active', true).order('confidence', { ascending: false }).limit(10),
    ]);

    return {
      profile: (profile.data ?? undefined) as AIProfile | undefined,
      latest_insights: (insights.data ?? []) as AIInsight[],
      coaching_cards: (coaching.data ?? []) as CoachingSession[],
      recommendations: (recommendations.data ?? []) as AIRecommendation[],
      detected_patterns: (patterns.data ?? []) as DetectedPattern[],
      recent_improvements: [],
      areas_to_improve: [],
    } as AIDashboardData;
  },
};

File: backend/src/api/routes/ai_foundation.py
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from src.api.deps import get_db, get_project_or_404
from src.models.project import Project
from src.services import ai_foundation
from src.schemas.ai_foundation import (
    AIProfileCreate, AIProfileUpdate,
    KnowledgeLinkCreate,
    AIProviderConfigCreate, AIProviderConfigUpdate,
    AIContextBuildRequest, CoachingSessionCreate,
)
from src.core.logging import get_logger

logger = get_logger(__name__)
router = APIRouter()


def _safe(fn, *args, **kwargs):
    try:
        return fn(*args, **kwargs)
    except Exception as exc:
        logger.error("AI Foundation query failed: %s", exc)
        raise HTTPException(status_code=500, detail="Failed to load AI data")


# ── Profile ──

@router.get("/profile")
def get_profile(project_id: UUID, db: Session = Depends(get_db)):
    _safe(ai_foundation.get_or_create_profile, db, project_id)
    return _safe(ai_foundation.get_or_create_profile, db, project_id)


@router.put("/profile")
def update_profile(project_id: UUID, body: AIProfileUpdate, db: Session = Depends(get_db)):
    return _safe(ai_foundation.update_profile, db, project_id, body.model_dump(exclude_unset=True))


@router.post("/profile/analyze")
def analyze_profile(project_id: UUID, db: Session = Depends(get_db)):
    return _safe(ai_foundation.analyze_trader_profile, db, project_id)


# ── Evaluations ──

@router.post("/evaluate/{trade_id}")
def evaluate_trade(project_id: UUID, trade_id: UUID, db: Session = Depends(get_db)):
    result = _safe(ai_foundation.evaluate_trade, db, project_id, trade_id)
    if not result:
        raise HTTPException(status_code=404, detail="Trade not found")
    return result


@router.get("/evaluations")
def list_evaluations(project_id: UUID, limit: int = 50, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_trade_evaluations, db, project_id, limit)


# ── Patterns ──

@router.post("/patterns/detect")
def detect_patterns(project_id: UUID, db: Session = Depends(get_db)):
    return _safe(ai_foundation.detect_patterns, db, project_id)


@router.get("/patterns")
def list_patterns(project_id: UUID, pattern_type: str = None, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_patterns, db, project_id, pattern_type)


# ── Knowledge Links ──

@router.post("/knowledge/links")
def create_link(project_id: UUID, body: KnowledgeLinkCreate, db: Session = Depends(get_db)):
    return _safe(ai_foundation.create_link, db, project_id, body.model_dump())


@router.get("/knowledge/links")
def list_links(project_id: UUID, entity_type: str = None, entity_id: UUID = None, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_links, db, project_id, entity_type, entity_id)


@router.delete("/knowledge/links/{link_id}")
def delete_link(project_id: UUID, link_id: UUID, db: Session = Depends(get_db)):
    if not _safe(ai_foundation.delete_link, db, link_id):
        raise HTTPException(status_code=404, detail="Link not found")
    return {"ok": True}


@router.post("/knowledge/auto-link")
def auto_link(project_id: UUID, db: Session = Depends(get_db)):
    count = _safe(ai_foundation.auto_link_trades, db, project_id)
    return {"linked": count}


@router.get("/knowledge/graph")
def get_graph(project_id: UUID, entity_type: str = None, entity_id: UUID = None, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_link_graph, db, project_id, entity_type, entity_id)


# ── Insights ──

@router.post("/insights/generate")
def generate_insights(project_id: UUID, db: Session = Depends(get_db)):
    return _safe(ai_foundation.generate_insights, db, project_id)


@router.get("/insights")
def list_insights(project_id: UUID, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_insights, db, project_id)


@router.put("/insights/{insight_id}/dismiss")
def dismiss_insight(project_id: UUID, insight_id: UUID, db: Session = Depends(get_db)):
    if not _safe(ai_foundation.dismiss_insight, db, insight_id):
        raise HTTPException(status_code=404, detail="Insight not found")
    return {"ok": True}


# ── Recommendations ──

@router.post("/recommendations/generate")
def generate_recommendations(project_id: UUID, db: Session = Depends(get_db)):
    return _safe(ai_foundation.generate_recommendations, db, project_id)


@router.get("/recommendations")
def list_recommendations(project_id: UUID, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_recommendations, db, project_id)


@router.put("/recommendations/{rec_id}/dismiss")
def dismiss_recommendation(project_id: UUID, rec_id: UUID, db: Session = Depends(get_db)):
    if not _safe(ai_foundation.dismiss_recommendation, db, rec_id):
        raise HTTPException(status_code=404, detail="Recommendation not found")
    return {"ok": True}


# ── Coaching ──

@router.post("/coaching/generate")
def generate_coaching(project_id: UUID, session_type: str = "daily", date: str = None, db: Session = Depends(get_db)):
    return _safe(ai_foundation.generate_coaching, db, project_id, session_type, date)


@router.get("/coaching")
def list_coaching(project_id: UUID, session_type: str = None, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_coaching_sessions, db, project_id, session_type)


# ── Summaries / Memory ──

@router.post("/summaries")
def create_summary(project_id: UUID, body: CoachingSessionCreate, db: Session = Depends(get_db)):
    return _safe(ai_foundation.create_summary, db, project_id, body.model_dump())


@router.get("/summaries")
def list_summaries(project_id: UUID, summary_type: str = None, period: str = None, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_summaries, db, project_id, summary_type, period)


@router.post("/summaries/performance")
def generate_performance_summary(project_id: UUID, db: Session = Depends(get_db)):
    return _safe(ai_foundation.generate_performance_summary, db, project_id)


# ── Context Builder ──

@router.post("/context")
def build_context(project_id: UUID, body: AIContextBuildRequest, db: Session = Depends(get_db)):
    return _safe(ai_foundation.build_context, db, project_id, body.trade_id, body.model_dump(exclude_unset=True))


# ── Providers ──

@router.get("/providers")
def list_providers(db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_providers, db)


@router.get("/providers/default")
def get_default_provider(db: Session = Depends(get_db)):
    result = _safe(ai_foundation.get_default_provider, db)
    if not result:
        raise HTTPException(status_code=404, detail="No default provider configured")
    return result


@router.post("/providers")
def create_provider(body: AIProviderConfigCreate, db: Session = Depends(get_db)):
    return _safe(ai_foundation.create_provider, db, body.model_dump())


@router.put("/providers/{provider_id}")
def update_provider(provider_id: UUID, body: AIProviderConfigUpdate, db: Session = Depends(get_db)):
    result = _safe(ai_foundation.update_provider, db, provider_id, body.model_dump(exclude_unset=True))
    if not result:
        raise HTTPException(status_code=404, detail="Provider not found")
    return result


@router.delete("/providers/{provider_id}")
def delete_provider(provider_id: UUID, db: Session = Depends(get_db)):
    if not _safe(ai_foundation.delete_provider, db, provider_id):
        raise HTTPException(status_code=404, detail="Provider not found")
    return {"ok": True}


# ── Dashboard ──

@router.get("/dashboard")
def get_ai_dashboard(project_id: UUID, db: Session = Depends(get_db)):
    return _safe(ai_foundation.get_ai_dashboard, db, project_id)
